### PR TITLE
Fix server oauth2 api request body www-form parameter decoding

### LINKED ISSUE #2484: Server /oauth2/token fails with POST application/x-www-form-urlencoded parameters
Compared to other oauth2 clients, the kontena-cli `Kontena::Client` sends the oauth2 token exchange request with a JSON body by default. It looks like the standard www-form request body decoding is broken:

```
POST /oauth2/token HTTP/1.1
Host: 46.101.140.29:80
User-Agent: Go-http-client/1.1
Content-Length: 67
Authorization: Basic ZDB...MGQ=
Content-Type: application/x-www-form-urlencoded
Accept-Encoding: gzip

code=tuW...H5&grant_type=authorization_code
```

```
HTTP/1.1 500 Internal Server Error
Content-Type: application/json
X-Kontena-Version: 1.3.0
Content-Length: 35

{"message":"Internal server error"}
```

```
TypeError (no implicit conversion of String into Integer):
  /app/app/routes/oauth2/token_api.rb:54:in `[]'
  /app/app/routes/oauth2/token_api.rb:54:in `block (2 levels) in <class:TokenApi>'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:713:in `always'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:448:in `is'
  /app/app/routes/oauth2/token_api.rb:46:in `block in <class:TokenApi>'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:250:in `instance_exec'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:250:in `block in call'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:248:in `catch'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:248:in `call'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda/plugins/error_handler.rb:52:in `call'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:222:in `block in build_rack_app'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:114:in `call'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:642:in `run'
  /app/server.rb:67:in `block (3 levels) in <class:Server>'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:771:in `if_match'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:494:in `on'
  /app/server.rb:66:in `block (2 levels) in <class:Server>'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:771:in `if_match'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:494:in `on'
  /app/server.rb:65:in `block in <class:Server>'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:250:in `instance_exec'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:250:in `block in call'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:248:in `catch'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:248:in `call'
  /usr/lib/ruby/gems/2.3.0/gems/roda-2.5.0/lib/roda.rb:222:in `block in build_rack_app'
  /app/app/middlewares/version_injector.rb:12:in `call'
  /app/app/middlewares/token_authentication.rb:38:in `call'
  /usr/lib/ruby/gems/2.3.0/gems/rack-1.6.4/lib/rack/static.rb:124:in `call'
  /usr/lib/ruby/gems/2.3.0/gems/rack-attack-5.0.1/lib/rack/attack.rb:147:in `call'
  /usr/lib/ruby/gems/2.3.0/gems/rack-1.6.4/lib/rack/commonlogger.rb:33:in `call'
  /app/app/middlewares/websocket_backend.rb:66:in `call'
  /usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/configuration.rb:225:in `call'
  /usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/server.rb:578:in `handle_request'
  /usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/server.rb:415:in `process_client'
  /usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/server.rb:275:in `block in run'
  /usr/lib/ruby/gems/2.3.0/gems/puma-3.6.0/lib/puma/thread_pool.rb:116:in `block in spawn_thread'

82.181.210.184 - - [15/Jun/2017:13:41:51 +0000] "POST /oauth2/token HTTP/1.1" 500 35 0.0075
```

Adding some debug logging to dump the params shows that the request body gets decoded as an `Array` of key-value pairs, rather than a `Hash`, which fails the `TypeError` when indexing with a `String`:

```
params=[["code", "tu...H5"], ["grant_type", "authorization_code"]]
```

This is because ruby [`URI.decode_www_form`](https://ruby-doc.org/stdlib-2.3.1/libdoc/uri/rdoc/URI.html#method-c-decode_www_form) does not return a `Hash`.

### PR DESCRIPTION
**Pull request by SpComb**
Fixes  #2484

### CODE REVIEW

File: server/app/routes/oauth2_api.rb [hunk 1]
```diff
@@ -24,7 +24,7 @@ def params_from_anywhere
        params =
          case request_content_type
          when JSON_MIME then JSON.parse(request.body.read)
-         when FORM_MIME then URI.decode_www_form(request.body.read)
+         when FORM_MIME then Hash[URI.decode_www_form(request.body.read)]
```
Comment (SpComb) [inline server/app/routes/oauth2_api.rb]: [`URI.decode_www_form`](https://ruby-doc.org/stdlib-2.3.1/libdoc/uri/rdoc/URI.html#method-c-decode_www_form) returns `Array` of `[[key, value]]`

Comment (kke) [inline server/app/routes/oauth2_api.rb]: to_h should work in that case too since it's on a fresh ruby.

**Comment by jakolehm** [comment]
Can we include this to 1.3.1?

**Comment by kke** [comment]
Yes, it's a bug and the fix is minimal.

It doesn't affect anyone unless they're building some custom api client, in which case they probably already used the JSON because form-data didn't work.

**Comment by SpComb** [comment]
:+1: for 1.3.1

**Pull request merged**